
## Shahwaiz
----
## Goal

Your goal in this part of assigment is to implement a Decision Tree Classifier for categorical variables.

**Note** Please note that you are allowed to use only those libraries which we have discussed in the class, i.e. numpy, scipy, pandas.

### Decision Tree Classifier

Now in this assignment we will be implementing the Decision Classifier for both Continuous and Categorical attributes.

Decision tree can be built by using any of the following split criterias, namely:
 - Information Gain
 - Gini Index
 - CART 

However, you are required here to implement the decision tree with information gain as splitting criterion.

Remember in my code i am not looking for maximizing the information gain, instead i am looking for minimizing the split entropy. Recall,
$$Information Gain  = H(D) - H(D_Y,D_N)$$

Where,

$H(D)$ is the data set entroy and $H(D_Y,D_N)$ is split entropy. Since $H(D)$ is constant for the given dataset so maximizing the entropy is equal to minimizing the split entropy and that is what is being represented in my code outputs.

In [2]:
%pylab inline
import scipy.stats
from collections import defaultdict  # default dictionary 
import copy

Populating the interactive namespace from numpy and matplotlib


In [3]:
#Cutomize the Matplotlib for beautiful plots...
# import dmStyle
# dmStyle.customize_mpl()


In [4]:
import pandas as pd
import tools as t # set of tools for plotting, data splitting, etc..

In [5]:
def getSplits(categories):
    '''
        function returns list of splits for the given list of categorical variables...
        
        Input:
        ------------
            categories: a list of unique categories...
        
        Return:
        ------------
            list of splits(tuples) for given list of categorical variables. Each pair of sublists
            defines the left and right splits, e.g. This list
            [('y', 'f'), ('s', 'g'), ('f'), ('y', 's', 'g')]
            
            defines two splits with each pair representing a different split.
        Examples:
        ------------
        splits=getSplits(['a1','a2','a3','a4']) will return 
        [('a1', 'a2', 'a4'), ('a3',), ('a2', 'a4'), 
        ('a1', 'a3'), ('a1', 'a4'), ('a3', 'a2'), ('a1', 'a3', 'a2'), 
        ('a4',), ('a3', 'a4'), ('a1', 'a2'), ('a1',), ('a3', 'a2', 'a4'), ('a2',), ('a1', 'a3', 'a4')]

            
    '''
    categories=set(categories)
    tsplits=t.get_powerset(categories,len(categories)-1) # get all the power sets with the given cardinality...
    flist=[]
    for s in tsplits:

        if not s in flist:
            r=categories.difference(s)
    #         print s, tuple(r)
            flist.append(s)
            flist.append(r)

    olist=[]

    for s in flist:
        ilist=[]

        for k in s:
            ilist.append(k)
        olist.append(tuple(ilist))

#     print olist
    
    return olist

In [6]:
splits=getSplits( [ 'a1','a2','a3','a4'] )
print splits,len(splits)

[('a1', 'a2', 'a4'), ('a3',), ('a2', 'a4'), ('a1', 'a3'), ('a1', 'a4'), ('a3', 'a2'), ('a1', 'a3', 'a2'), ('a4',), ('a3', 'a4'), ('a1', 'a2'), ('a1',), ('a3', 'a2', 'a4'), ('a2',), ('a1', 'a3', 'a4')] 14


In [7]:
class Node:
    def __init__(self,purity,klasslabel='',score=0,split=[],fidx=-1):
        '''
            purity: purity level at which to stop
            klasslabel: klasslabel of the node, (for leaf node)
            score: information gain of the newly added node
            split: splitting threshold
            fidx: feature index            
        '''
        self.lchild=None
        self.rchild=None
        self.klasslabel=klasslabel        
        self.split=split
        self.score=score
        self.fidx=fidx
        self.purity=purity
        self.ftype= 'categorical' if type(self.split) in [tuple, str, numpy.string_] else 'continuous' # feature type 

        
    def set_childs(self,lchild,rchild):
        # YOUR CODE HERE
        raise NotImplementedError()
    def isleaf(self):
        # YOUR CODE HERE
        if(self.lchild is  None and self.rchild is  None):
            return True
        else:
            return False
        #raise NotImplementedError()
    def isless_than_eq(self, X):
        # YOUR CODE HERE
        raise NotImplementedError()
    def get_str(self):        
        if self.isleaf():
            return 'C(class={},Purity={})'.format(self.klasslabel,self.purity)
        else:
            return 'I(Fidx={},Score={},Split={})'.format(self.fidx,self.score,self.split)
    

In [8]:
# A placeholder class 
# TODO: You have to implement the following class, remember from the lectures that you will 
# need to build a model for each different class you are trying to identify...

In [115]:
### import pdb
## Your code goes here...
# You might need to define auxliary classes for composition.. ?
#DecisionTree
class DecisionTree:
    ''' Implements the Decision Tree For Classification... '''
    def __init__(self, purityp, exthreshold, maxdepth=10):        
        self.purity = purityp
        self.exthreshold = exthreshold
        self.maxdepth = maxdepth
        self.root=None
        pass
    def train(self, X, Y):
        ''' Train Decision Tree using the given 
            X [m x d] data matrix and Y labels matrix
            
            Input:
            ------
            X: [m x d] a data matrix of m d-dimensional examples.
            Y: [m x 1] a label vector.
            
            Returns:
            -----------
            Nothing
            '''
        #nexamples,nfeatures=X.shape
        ## now go and train a model for each class...
        # YOUR CODE HERE
        #Xc=copy.copy(X)
        #Yc=copy.copy(Y)
        self.build_tree(X,Y,3)
        return 
       
    def build_tree(self, X, Y, depth,currNode=None):
            """ 
                Function is used to recursively build the decision Tree 

                Input
                -----
                X: [m x d] a data matrix of m d-dimensional examples.
                Y: [m x 1] a label vector.

                Returns
                -------
                root node of the built tree...
            """
            
            if(X.size==0):
                currNode=None
                return 
            
            #print "Y::",np.count_nonzero(Y=='f')
            #print "X::",X
            nexamples, nfeatures=X.shape
            #if(self.find_depth()>=10 ):
                #return self.root     
            #if(Y.size==0):
                #return 
            klasses=np.unique(Y)
            # YOUR CODE HERE
            p_arr=[]
            argmax_flag=False
            for i in klasses:
                    argmax_flag=True
                    p_arr.append( X[Y==i].shape[0])#/float(nexamples) )
            
            p_idx=np.argmax(p_arr)
            purityD=p_arr[p_idx]/float(nexamples)
            #print "purityD::",purityD
            if (nexamples<=self.exthreshold or purityD>=self.purity or self.find_depth()>6):
                
                currNode.purity=purityD
                currNode.klasslabel=klasses[p_idx]
                
                
                currNode.lchild=None
                currNode.rchild=None
                return self.root    
            split_point=None
            score=100
            
            Xlidx=[]
            Xridx=[]
            bestFeat=-1
            
            #X=X.T
            for feat in range(X.shape[1]):
                #def evaluate_numerical_attribute(self,feat, Y)
                #print "feat:",feat
                split,mingain_score,tmpXlidx,tmpXridx = self.evaluate_categorical_attribute(X[:,feat], Y)
                if(mingain_score < score):  
                    #print "split:",split
                    split_point=split
                    score=mingain_score
                    Xlidx=tmpXlidx
                    Xridx=tmpXridx
                    bestFeat=feat
            ###now we have the best split point for best feature
            ###split the X into two parts Dy and Dn on basis of split point
            ###Book:// partition D into DY and DN using split point∗, and call recursively 
            
            #print "split_store:",bestFeat,split_point
            DY_exp=X[ Xlidx ]
            DN_exp=X[ Xridx ]
            Y_l=Y[ Xlidx ]
            Y_r=Y[ Xridx ]
            #print "Xlidx:",Xlidx
            #print "Xridx:",Xridx
            
            ###***********************************************************###
            #################################################################
            ###def __init__(self,purity,klasslabel='',score=0,split=[],fidx=-1):
            
            split_store=[ bestFeat,split_point]
            #print "split_store::",split_store
            if(self.root==None):
                
                self.root=Node(purityD,'',score,split_store,-1)
                self.root.lchild=Node(-1,'noclass',-1,[-1,-1],-1)
                self.root.rchild=Node(-1,'noclass',-1,[-1,-1],-1)
                
                self.build_tree(DY_exp,Y_l,10,self.root.lchild)
                self.build_tree(DN_exp,Y_r,10,self.root.rchild)
                
            else:
                
                currNode.purity=purityD
                currNode.klasslabel=''
                currNode.score=score
                currNode.split=split_store
                currNode.fidx=-1
                
                currNode.lchild=Node(-1,'noclass',-1,[-1,-1],-1)
                currNode.rchild=Node(-1,'noclass',-1,[-1,-1],-1)
                
                self.build_tree(DY_exp,Y_l,10,currNode.lchild)
                self.build_tree(DN_exp,Y_r,10,currNode.rchild)
                
            return
        
    def test(self, X):
        
        ''' Test the trained classifiers on the given set of examples 
        
                   
            Input:
            ------
            X: [m x d] a data matrix of m d-dimensional test examples.
           
            Returns:
            -----------
                pclass: the predicted class for each example, i.e. to which it belongs
        '''
        
        nexamples, nfeatures = X.shape
        pclasses = self.predict(X)
        
        # your code go here...
        
        return np.array(pclasses)
    def evaluate_categorical_attribute(self, feat, Y):
            """
            Evaluates the cateogrical attribute for all possible split points, i.e. 2^(m-1)-2 for
            possible feature selection

            Input:
            ---------
            feat: a categorical feature
            Y: labels

            Returns:
            ----------
            v: splitting threshold
            score: splitting score
            Xlidx: Index of examples belonging to left child node
            Xridx: Index of examples belonging to right child node

            """
            categories = set(feat)
            splits = getSplits(categories) if len(categories) > 1 else tuple(categories)
            #freq = scipy.stats.itemfreq(Y)
            ############################################
            #******************************************#
            # YOUR CODE HERE
            ###calcculate the entropy and return it
            klasses=np.unique(Y)
            splits_entrop=[]
            
            sz_tot=float(Y.shape[0])   
            for s in splits:
                    idx=np.in1d(feat,s).nonzero()[0]
                    idxv=np.in1d(feat,s,invert=True).nonzero()[0]
                    
                    s_y = Y[idx]
                    s_n = Y[idxv]
                    
                    p_y=s_y.shape[0] / sz_tot
                    p_n=s_n.shape[0] / sz_tot
                    
                    
                    size_y=float(s_y.shape[0])
                    size_n=float(s_n.shape[0])
                    
                    h_d_x=0.0
                    h_d_n=0.0
                        
                    for k in klasses:
                        
                        p = s_y[s_y==k].shape[0]/size_y
                        
                        tmp=p * np.log2( p + np.spacing(1) )
                        
                        h_d_x = h_d_x + tmp
             
                        if(size_n!=0):
                            p2=s_n[s_n==k].shape[0]/size_n
                        else:
                            p2=0.0
                        h_d_n=  h_d_n + p2 * np.log2( p2 + np.spacing(1) )
                    
                    h_d_n=h_d_n*-1
                    h_d_x=h_d_x*-1
                    #Now cal the Entropy for (s,Universe-s) # s is from splits array
                    entrop= (p_n * h_d_n) + (p_y * h_d_x)
                    #print "entrop::",entrop
                    if(entrop!=0 and entrop is not nan):
                        splits_entrop.append(entrop)    
                    else:
                        splits_entrop.append(100)
                        
            ###now we have the min entrop and the split point which gives the min entrop 
            
            bidx=np.argmin(splits_entrop)
            csplit=splits[bidx]
            mingain=splits_entrop[bidx]
            
            Xlidx=np.in1d(feat,csplit).nonzero()[0]#points that satisfy the split point 
            Xridx=np.in1d(feat,csplit,invert=True).nonzero()[0]        
 
            
            #******************************************#
            ############################################
            
            #raise NotImplementedError()
            return csplit, mingain, Xlidx, Xridx
        
    def evaluate_numerical_attribute(self, feat, Y):
        '''
            Evaluates the numerical attribute for all possible split points for
            possible feature selection
            
            Input:
            ---------
            feat: a contiuous feature
            Y: labels
            
            Returns:
            ----------
            v: splitting threshold
            score: splitting score
            Xlidx: Index of examples belonging to left child node
            Xridx: Index of examples belonging to right child node
            
        '''
        
        classes = np.unique(Y)
        nclasses = len(self.classes)
        f = np.sort(feat)
        
        # YOUR CODE HERE
        raise NotImplementedError()
        
        return split, mingain, Xlidx, Xridx
    def predict(self, X):
        
        """
        Test the trained classifiers on the given example X
        
                   
            Input:
            ------
            X: [1 x d] a d-dimensional test example.
           
            Returns:
            -----------
                pclass: the predicted class for the given example, i.e. to which it belongs
        """
        z = []
        
        for idx in range(X.shape[0]):
            
            z.append(self._predict(self.root, X[idx, :]))
        
        return z 
    
    def _predict(self, node, X):
        #[ bestFeat,split_point]
        # YOUR CODE HERE
        if( node.lchild == None or node.rchild == None):
            return node.klasslabel
        else:
            bFeat=node.split[0]
            val=node.split[1]
            
            #Xlidx=np.in1d(X , csplit).nonzero()[0]#points that satisfy the split point 
            
            #Xridx=np.in1d(feat,csplit,invert=True).nonzero()[0]        
            l_flag=False
            for i in val:
                if(i==X[bFeat]):
                    l_flag=True
            if( l_flag ):
                return self._predict(node.lchild,X)
            else:
                return self._predict(node.rchild,X)
        
        
        #raise NotImplementedError()

    def __str__(self):
        str = '---------------------------------------------------'
        str += '\n A Decision Tree With Depth={}'.format(self.find_depth())
        str += self.__print(self.root)
        str += '\n---------------------------------------------------'
        return str  # self.__print(self.tree)        
        
     
   
    def find_depth(self):
        return self._find_depth(self.root)
    def _find_depth(self, node):
        if not node:
            return
        if node.isleaf():
            return 1
        else:
            return max(self._find_depth(node.lchild), self._find_depth(node.rchild)) + 1
    def __print(self, node, depth=0):
        
        ret = ""

        # Print right branch
        if node.rchild:
            ret += self.__print(node.rchild, depth + 1)

        # Print own value
        
        ret += "\n" + ("    "*depth) + node.get_str()

        # Print left branch
        if node.lchild:
            ret += self.__print(node.lchild, depth + 1)
        
        return ret

In [77]:
# Lets train a Decision Tree Classifier two features
#%pdb
feat=np.arange(2)
dt=DecisionTree(0.95,5)
dt.train(Xtrain[:,feat],Ytrain)

# Lets test our code for the given example in the book.

In [164]:
#%pdb off
#dt=DecisionTree(0.95,5,5)
split, gain, Xlidx, Xridx = evaluate_categorical_attribute(tx,Y)

#q,e,r,p=evaluate_categorical_attribute(tx,Y)

print 
print split, gain
#Automatic pdb calling has been turned OFF
#('a1',) 0.510702350316
#h_d_y=0.991
#h_d_n=0.673



('a1',) 0.510702350316


In [11]:
#load the Iris dataset
tdata=pd.read_csv('./iris.data')
tdata.columns=['SepalLength','SepalWidth','PetalLength','PetalWidth','Class']

In [68]:
# Get your data in matrix
X=np.asarray(data.ix[:,1:].dropna())
Y=np.asarray(data.ix[:,0].dropna())
print " Data Set Dimensions=", X.shape, " True Class labels dimensions", Y.shape   
# Split your data into training and test-set... 
# see the documentation of split_data in tools for further information...
Xtrain,Ytrain,Xtest,Ytest=t.split_data(X,Y)

print " Training Data Set Dimensions=", Xtrain.shape, "Training True Class labels dimensions", Ytrain.shape   
print " Test Data Set Dimensions=", Xtest.shape, "Test True Class labels dimensions", Ytrain.shape   


from nose.tools import assert_greater_equal
Xtrain,Ytrain,Xtest,Ytest=t.split_data(X,Y)
feat=np.arange(2)
dt=DecisionTree(0.95,5,10)
dt.train(Xtrain[:,feat],Ytrain)


pclasses=dt.predict(Xtest[:,feat])
acc = np.sum(pclasses==Ytest)/float(Ytest.shape[0])

assert_greater_equal(acc, 0.60)

 Data Set Dimensions= (8124L, 22L)  True Class labels dimensions (8124L,)
 Training Data Set Dimensions= (5687L, 22L) Training True Class labels dimensions (5687L,)
 Test Data Set Dimensions= (2437L, 22L) Test True Class labels dimensions (5687L,)


In [13]:
tx=tdata['SepalLength'].dropna()

tx[(tdata['SepalLength']>=4.3) & (tdata['SepalLength']<=5.2)]='a1'
tx[(tdata['SepalLength']>5.2) & (tdata['SepalLength']<=6.1)]='a2'
tx[(tdata['SepalLength']>6.1) & (tdata['SepalLength']<=7.0)]='a3'
tx[(tdata['SepalLength']>7.0) & (tdata['SepalLength']<=7.9)]='a4'

In [14]:
print tx.values
Y=tdata['Class'].dropna()
Y[Y=='Iris-virginica']='Iris-versicolor'
Y=Y.values
print Y

['a1' 'a1' 'a1' 'a1' 'a2' 'a1' 'a1' 'a1' 'a1' 'a2' 'a1' 'a1' 'a1' 'a2' 'a2'
 'a2' 'a1' 'a2' 'a1' 'a2' 'a1' 'a1' 'a1' 'a1' 'a1' 'a1' 'a1' 'a1' 'a1' 'a1'
 'a2' 'a1' 'a2' 'a1' 'a1' 'a2' 'a1' 'a1' 'a1' 'a1' 'a1' 'a1' 'a1' 'a1' 'a1'
 'a1' 'a1' 'a2' 'a1' 'a3' 'a3' 'a3' 'a2' 'a3' 'a2' 'a3' 'a1' 'a3' 'a1' 'a1'
 'a2' 'a2' 'a2' 'a2' 'a3' 'a2' 'a2' 'a3' 'a2' 'a2' 'a2' 'a3' 'a2' 'a3' 'a3'
 'a3' 'a3' 'a2' 'a2' 'a2' 'a2' 'a2' 'a2' 'a2' 'a2' 'a3' 'a3' 'a2' 'a2' 'a2'
 'a2' 'a2' 'a1' 'a2' 'a2' 'a2' 'a3' 'a1' 'a2' 'a3' 'a2' 'a4' 'a3' 'a3' 'a4'
 'a1' 'a4' 'a3' 'a4' 'a3' 'a3' 'a3' 'a2' 'a2' 'a3' 'a3' 'a4' 'a4' 'a2' 'a3'
 'a2' 'a4' 'a3' 'a3' 'a4' 'a3' 'a2' 'a3' 'a4' 'a4' 'a4' 'a3' 'a3' 'a2' 'a4'
 'a3' 'a3' 'a2' 'a3' 'a3' 'a3' 'a2' 'a3' 'a3' 'a3' 'a3' 'a3' 'a3' 'a2']
['Iris-setosa' 'Iris-setosa' 'Iris-setosa' 'Iris-setosa' 'Iris-setosa'
 'Iris-setosa' 'Iris-setosa' 'Iris-setosa' 'Iris-setosa' 'Iris-setosa'
 'Iris-setosa' 'Iris-setosa' 'Iris-setosa' 'Iris-setosa' 'Iris-setosa'
 'Iris-setosa' 'Iris-setosa' 'I

In [92]:
categories = set(tx)
print categories

set(['a1', 'a3', 'a2', 'a4'])


In [167]:
#%pdb off
dt=DecisionTree(0.95,5,5)
split, gain, Xlidx, Xridx = dt.evaluate_categorical_attribute(tx,Y)

#q,e,r,p=evaluate_categorical_attribute(tx,Y)

print 
print split, gain
#Automatic pdb calling has been turned OFF
#('a1',) 0.510702350316


('a1',) 0.510702350316


In [168]:
from nose.tools import assert_almost_equal, assert_almost_equals

%pdb off
dt=DecisionTree(0.95,5,5)
split, gain, Xlidx, Xridx = dt.evaluate_categorical_attribute(tx,Y)


assert_almost_equal('a1', split[0])
assert_almost_equal(gain, 0.51, 1)

Automatic pdb calling has been turned OFF


# Now lets test on [Mushroom](https://archive.ics.uci.edu/ml/machine-learning-databases/mushroom/agaricus-lepiota.names) dataset


This data set includes descriptions of hypothetical samples corresponding to 23 species of gilled mushrooms in the Agaricus and Lepiota Family (pp. 500-525).  Each species is identified as definitely edible, definitely poisonous, or of unknown edibility and not recommended.  This latter class was combined with the poisonous one.  The Guide clearly states that there is no simple rule for determining the edibility of a mushroom; no rule like ``leaflets three, let it be'' for Poisonous Oak and Ivy.


- Number of Instances: 8124

- Number of Attributes: 22 (all nominally valued)

- Attribute Information: (classes: edible=e, poisonous=p)
         1. cap-shape:                bell=b,conical=c,convex=x,flat=f, knobbed=k,sunken=s
         2. cap-surface:              fibrous=f,grooves=g,scaly=y,smooth=s
         3. cap-color:                brown=n,buff=b,cinnamon=c,gray=g,green=r, pink=p,purple=u,red=e,white=w,yellow=y
         4. bruises?:                 bruises=t,no=f
         5. odor:                     almond=a,anise=l,creosote=c,fishy=y,foul=f, musty=m,none=n,pungent=p,spicy=s
         6. gill-attachment:          attached=a,descending=d,free=f,notched=n
         7. gill-spacing:             close=c,crowded=w,distant=d
         8. gill-size:                broad=b,narrow=n
         9. gill-color:               black=k,brown=n,buff=b,chocolate=h,gray=g, green=r,orange=o,pink=p,purple=u,red=e,white=w,yellow=y
        10. stalk-shape:              enlarging=e,tapering=t
        11. stalk-root:               bulbous=b,club=c,cup=u,equal=e, rhizomorphs=z,rooted=r,missing=?
        12. stalk-surface-above-ring: fibrous=f,scaly=y,silky=k,smooth=s
        13. stalk-surface-below-ring: fibrous=f,scaly=y,silky=k,smooth=s
        14. stalk-color-above-ring:   brown=n,buff=b,cinnamon=c,gray=g,orange=o, pink=p,red=e,white=w,yellow=y
        15. stalk-color-below-ring:   brown=n,buff=b,cinnamon=c,gray=g,orange=o, pink=p,red=e,white=w,yellow=y
        16. veil-type:                partial=p,universal=u
        17. veil-color:               brown=n,orange=o,white=w,yellow=y
        18. ring-number:              none=n,one=o,two=t
        19. ring-type:                cobwebby=c,evanescent=e,flaring=f,large=l, none=n,pendant=p,sheathing=s,zone=z
        20. spore-print-color:        black=k,brown=n,buff=b,chocolate=h,green=r, orange=o,purple=u,white=w,yellow=y
        21. population:               abundant=a,clustered=c,numerous=n, scattered=s,several=v,solitary=y
        22. habitat:                  grasses=g,leaves=l,meadows=m,paths=p, urban=u,waste=w,woods=d

- Missing Attribute Values: 2480 of them (denoted by "?"), all for  attribute #11.

- Class Distribution: 
        --    edible: 4208 (51.8%)
        -- poisonous: 3916 (48.2%)
        --     total: 8124 instances

In [15]:
#load the data set
data=pd.read_csv('./mushrooms.csv')
data.columns=columns=['class', 'cap-shape','cap-surface','cap-color','bruises?','odor','gill-attachment','gill-spacing','gill-size','gill-color','stalk-shape','stalk-root','stalk-surface-above-ring','stalk-surface-below-ring','stalk-color-above-ring','stalk-color-below-ring','veil-type','veil-color','ring-number','ring-type','spore-print-color','population','habitat']
print data.describe()

       class cap-shape cap-surface cap-color bruises?  odor gill-attachment  \
count   8124      8124        8124      8124     8124  8124            8124   
unique     2         6           4        10        2     9               2   
top        e         x           y         n        f     n               f   
freq    4208      3656        3244      2284     4748  3528            7914   

       gill-spacing gill-size gill-color   ...   stalk-surface-below-ring  \
count          8124      8124       8124   ...                       8124   
unique            2         2         12   ...                          4   
top               c         b          b   ...                          s   
freq           6812      5612       1728   ...                       4936   

       stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
count                    8124                   8124      8124       8124   
unique                      9                      9         1  

In [16]:
data.tail()

class cap-shape cap-surface cap-color bruises? odor gill-attachment  \
8119     e         k           s         n        f    n               a   
8120     e         x           s         n        f    n               a   
8121     e         f           s         n        f    n               a   
8122     p         k           y         n        f    y               f   
8123     e         x           s         n        f    n               a   

     gill-spacing gill-size gill-color   ...   stalk-surface-below-ring  \
8119            c         b          y   ...                          s   
8120            c         b          y   ...                          s   
8121            c         b          n   ...                          s   
8122            c         n          b   ...                          k   
8123            c         b          y   ...                          s   

     stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
8119                      o                      o         p          o   
8120                      o                      o         p          n   
8121                      o                      o         p          o   
8122                      w                      w         p          w   
8123                      o                      o         p          o   

     ring-number ring-type spore-print-color population habitat  
8119           o         p                 b          c       l  
8120           o         p                 b          v       l  
8121           o         p                 b          c       l  
8122           o         e                 w          v       l  
8123           o         p                 o          c       l  

[5 rows x 23 columns]

In [131]:
# Get your data in matrix
X=np.asarray(data.ix[:,1:].dropna())
Y=np.asarray(data.ix[:,0].dropna())
print " Data Set Dimensions=", X.shape, " True Class labels dimensions", Y.shape   


 Data Set Dimensions= (8124L, 22L)  True Class labels dimensions (8124L,)


In [132]:
# Split your data into training and test-set... 
# see the documentation of split_data in tools for further information...
Xtrain,Ytrain,Xtest,Ytest=t.split_data(X,Y)

print " Training Data Set Dimensions=", Xtrain.shape, "Training True Class labels dimensions", Ytrain.shape   
print " Test Data Set Dimensions=", Xtest.shape, "Test True Class labels dimensions", Ytrain.shape   


 Training Data Set Dimensions= (5687L, 22L) Training True Class labels dimensions (5687L,)
 Test Data Set Dimensions= (2437L, 22L) Test True Class labels dimensions (5687L,)


In [69]:
# Lets train a Decision Tree Classifier two features
#%pdb
feat=np.arange(2)
dt=DecisionTree(0.95,5)
dt.train(Xtrain[:,feat],Ytrain)

In [70]:
print dt

---------------------------------------------------
 A Decision Tree With Depth=7
    C(class=e,Purity=0.896551724138)
I(Fidx=-1,Score=0.968050015674,Split=[0, ('x', 'c', 'k', 'f')])
        C(class=p,Purity=0.575725593668)
    I(Fidx=-1,Score=0.965735742057,Split=[1, ('f',)])
            C(class=e,Purity=0.655607166556)
        I(Fidx=-1,Score=0.916074545985,Split=[0, ('k',)])
                C(class=noclass,Purity=-1)
            I(Fidx=-1,Score=0.453716339187,Split=[0, 'k'])
                    C(class=noclass,Purity=-1)
                I(Fidx=-1,Score=0.453716339187,Split=[0, 'k'])
                        C(class=noclass,Purity=-1)
                    I(Fidx=-1,Score=0.453716339187,Split=[0, 'k'])
                        C(class=e,Purity=0.904761904762)
---------------------------------------------------


In [157]:
pclasses=dt.predict(Xtest[:,feat])
#Lets see how good we are doing, by finding the accuracy on the test set..
print np.sum(pclasses==Ytest)
print "Accuracy = ", np.sum(pclasses==Ytest)/float(Ytest.shape[0])

2392
Accuracy =  0.981534673779


In [140]:
from nose.tools import assert_greater_equal
Xtrain,Ytrain,Xtest,Ytest=t.split_data(X,Y)
feat=np.arange(2)
dt=DecisionTree(0.95,5,10)
dt.train(Xtrain[:,feat],Ytrain)


pclasses=dt.predict(Xtest[:,feat])
acc = np.sum(pclasses==Ytest)/float(Ytest.shape[0])

assert_greater_equal(acc, 0.60)

In [141]:
from nose.tools import assert_greater_equal
Xtrain,Ytrain,Xtest,Ytest=t.split_data(X,Y)

dt=DecisionTree(0.95,5,10)
dt.train(Xtrain,Ytrain)

pclasses=dt.predict(Xtest)
acc = np.sum(pclasses==Ytest)/float(Ytest.shape[0])

assert_greater_equal(acc, 0.90)

# Lets Train on all the features and for both the classes....

In [27]:
# Split your data into training and test-set... 
# see the documentation of split_data in tools for further information...
Xtrain,Ytrain,Xtest,Ytest=t.split_data(X,Y)

print " Training Data Set Dimensions=", Xtrain.shape, "Training True Class labels dimensions", Ytrain.shape   
print " Test Data Set Dimensions=", Xtest.shape, "Test True Class labels dimensions", Ytrain.shape   


 Training Data Set Dimensions= (5687L, 22L) Training True Class labels dimensions (5687L,)
 Test Data Set Dimensions= (2437L, 22L) Test True Class labels dimensions (5687L,)


In [155]:
feat=arange(22)#[0, 1, 2, 3]
dt=DecisionTree(0.95,5,10)
dt.train(Xtrain[:,feat],Ytrain)
pclasses=dt.predict(Xtest[:,feat])
#Lets see how good we are doing, by finding the accuracy on the test set..
print np.sum(pclasses==Ytest)
print "Accuracy = ", np.sum(pclasses==Ytest)/float(Ytest.shape[0])

2392
Accuracy =  0.981534673779


In [75]:
print dt

---------------------------------------------------
 A Decision Tree With Depth=2
    C(class=p,Purity=1.0)
I(Fidx=-1,Score=0.0955909088803,Split=[4, ('a', 'l', 'n')])
    C(class=e,Purity=0.972919418758)
---------------------------------------------------


What can you conclude ?
====================
Please write your observation....



# Cross-Validation

Until now we have been splitting the dataset into a training and test set rather randomly and were reporting a rather artifical performance. Now we are going to test our system exhaustively by making use of k-fold [cross validation](http://en.wikipedia.org/wiki/Cross-validation_%28statistics%29). 

Now go and tune your hyper-parameters (purity, exthreshold) to opitmize the performance.

In [76]:
# Now lets cross validate for best paramters, and test the result...
# We will be training four different models on four different partitions of data set and 
# then will be reporting the mean accuracy of the four classifiers.

nfolds=4 # lets use four folds..
folds=t.generate_folds(X,Y,nfolds)
features=arange(22)#[0, 1, 2, 3] # features to use for our system
#now lets train and test on these folds...
totacc=[]
for k in range(nfolds):
    dt=DecisionTree(0.95,5,7)
    dt.train(folds[k][0][:,features],folds[k][1])
    pclasses=dt.predict(folds[k][2][:,features])
    acc=np.sum(pclasses==folds[k][3])/float(folds[k][3].shape[0])
    print dt
    print "[Info] Fold {} Accuracy = {}".format(k+1, acc)
    totacc.append(acc)

print totacc, '\n Mean Accuracy =', np.mean(totacc)

Generating CV data for 2 classes
---------------------------------------------------
 A Decision Tree With Depth=2
    C(class=p,Purity=1.0)
I(Fidx=-1,Score=0.0974165640063,Split=[4, ('a', 'l', 'n')])
    C(class=e,Purity=0.972273567468)
---------------------------------------------------
[Info] Fold 1 Accuracy = 0.985228951256
---------------------------------------------------
 A Decision Tree With Depth=2
    C(class=p,Purity=1.0)
I(Fidx=-1,Score=0.0965663396895,Split=[4, ('a', 'l', 'n')])
    C(class=e,Purity=0.972573189522)
---------------------------------------------------
[Info] Fold 2 Accuracy = 0.984736582964
---------------------------------------------------
 A Decision Tree With Depth=2
    C(class=p,Purity=1.0)
I(Fidx=-1,Score=0.098264230335,Split=[4, ('a', 'l', 'n')])
    C(class=e,Purity=0.971974129966)
---------------------------------------------------
[Info] Fold 3 Accuracy = 0.985721319547
---------------------------------------------------
 A Decision Tree With Dep

In [78]:
# Now lets cross validate for best paramters, and test the result...
# We will be training four different models on four different partitions of data set and 
# then will be reporting the mean accuracy of the four classifiers.

nfolds=4 # lets use four folds..
folds=t.generate_folds(X,Y,nfolds)
features=[0,1, 2, 3] # features to use for our system
#now lets train and test on these folds...

#Lets perform the grid search...
purity=np.linspace(0.85,0.97,13)
nexamp=np.linspace(5,25,21)  

params=np.zeros((len(purity),len(nexamp)))
                   
for p in range(len(purity)):
    for n in range(len(nexamp)):
        totacc=[]
        for k in range(nfolds):
            dt=DecisionTree(purity[p],nexamp[n],7)#purity[p],nexamp[n])
            dt.train(folds[k][0][:,features],folds[k][1])
            pclasses=dt.predict(folds[k][2][:,features])
            acc=np.sum(pclasses==folds[k][3])/float(folds[k][3].shape[0])
            print "[Info] Fold {} Accuracy = {}".format(k+1, acc)
            totacc.append(acc)
        params[p,n]=np.mean(totacc)
        print totacc, '\nPurity={}, Nexample-threshold={}, Mean Accuracy ={}'.format(purity[p],nexamp[n], np.mean(totacc))

Generating CV data for 2 classes
[Info] Fold 1 Accuracy = 0.80059084195
[Info] Fold 2 Accuracy = 0.81388478582
[Info] Fold 3 Accuracy = 0.813392417528
[Info] Fold 4 Accuracy = 0.807976366322
[0.80059084194977848, 0.81388478581979318, 0.81339241752831115, 0.80797636632200887] 
Purity=0.85, Nexample-threshold=5.0, Mean Accuracy =0.808961102905
[Info] Fold 1 Accuracy = 0.80059084195
[Info] Fold 2 Accuracy = 0.81388478582
[Info] Fold 3 Accuracy = 0.813392417528
[Info] Fold 4 Accuracy = 0.807976366322
[0.80059084194977848, 0.81388478581979318, 0.81339241752831115, 0.80797636632200887] 
Purity=0.85, Nexample-threshold=6.0, Mean Accuracy =0.808961102905
[Info] Fold 1 Accuracy = 0.80059084195
[Info] Fold 2 Accuracy = 0.81388478582
[Info] Fold 3 Accuracy = 0.813392417528
[Info] Fold 4 Accuracy = 0.807976366322
[0.80059084194977848, 0.81388478581979318, 0.81339241752831115, 0.80797636632200887] 
Purity=0.85, Nexample-threshold=7.0, Mean Accuracy =0.808961102905
[Info] Fold 1 Accuracy = 0.8005908

KeyboardInterrupt: 

In [32]:
np.unravel_index(np.argmax(params), params.shape)

(0, 0)